File: genetic_algorithm.py
import random
import time

# Genera la población inicial usando PERMUTACIONES exclusivas
def generate_initial_population(population_size: int, n_queens: int) -> list:
    population = []
    for _ in range(population_size):
        # Al usar una permutación aseguramos que no haya colisiones horizontales desde el día 1
        individual = list(range(n_queens))
        random.shuffle(individual)
        population.append(individual)
    
    return population

# No va a haber colisiones verticales porque el individuo tiene un valor por columna.
def calculate_fitness(individual: list) -> float:
    n = len(individual)
    max_fitness = (n * (n - 1)) / 2
    
    # Ya no hay colisiones horizontales (ni verticales) gracias al uso estricto de permutaciones.
    # Eliminar este cálculo reduce enormemente el tiempo de procesamiento por generación.
    left_diagonal = [0] * (2 * n)
    right_diagonal = [0] * (2 * n)
    for i in range(n):
        left_diagonal[i + individual[i] - 1] += 1
        right_diagonal[n - i + individual[i] - 2] += 1

    diagonal_collisions = 0
    for i in range(2 * n - 1):
        counter = 0
        if left_diagonal[i] > 1:
            counter += left_diagonal[i] - 1
        if right_diagonal[i] > 1:
            counter += right_diagonal[i] - 1
        diagonal_collisions += counter / (n - abs(i - n + 1))
    
    fitness_val = int(max_fitness - diagonal_collisions)
    fitness_score = fitness_val / max_fitness
    
    return fitness_score

# Selecciona los individuos usando Torneo (Tournament Selection) para mantener la presión selectiva
def select_fitted_population(population: list, fitness_scores: list) -> list:
    parents = []
    pop_size = len(population)
    tournament_size = 5 # Seleccionamos al mejor de 5 individuos al azar
    for _ in range(pop_size):
        best_idx = random.randint(0, pop_size - 1)
        for _ in range(tournament_size - 1):
            idx = random.randint(0, pop_size - 1)
            if fitness_scores[idx] > fitness_scores[best_idx]:
                best_idx = idx
        parents.append(population[best_idx])
    return parents

# Realiza el Cruce de Orden (Order Crossover - OX1) para preservar las permutaciones.
def crossover(parent1: list, parent2: list) -> tuple:
    n = len(parent1)
    # Elegir dos puntos de corte aleatorios
    start, end = sorted(random.sample(range(n), 2))
    
    def apply_ox(p1, p2):
        offspring = [-1] * n
        # Copiar segmento del padre 1
        offspring[start:end+1] = p1[start:end+1]
        
        # Rellenar con los del padre 2 omitiendo los repetidos
        p2_idx = (end + 1) % n
        off_idx = (end + 1) % n
        
        while -1 in offspring:
            gene = p2[p2_idx]
            if gene not in offspring:
                offspring[off_idx] = gene
                off_idx = (off_idx + 1) % n
            p2_idx = (p2_idx + 1) % n
            
        return offspring

    return apply_ox(parent1, parent2), apply_ox(parent2, parent1)

# Mutación por Intercambio (Swap Mutation) para no romper las permutaciones.
def mutate(individual: list, mutation_rate: float) -> list:
    n = len(individual)
    for i in range(n):
        if random.random() <= mutation_rate:
            # En lugar de asignar un valor al azar (que crearía duplicados), intercambiamos
            swap_idx = random.randint(0, n - 1)
            individual[i], individual[swap_idx] = individual[swap_idx], individual[i]
            
    return individual

# Visualiza el tablero.
def visualize_board(individual: list):
    n = len(individual)
    print(f"\nSolución encontrada: {individual}")
    for row in range(n):
        row_str = ""
        for col in range(n):
            if individual[col] == row:
                row_str += "Q "
            else:
                row_str += ". "
        print(row_str)
    print()

def run_genetic_algorithm(n_queens: int = 32, population_size: int = 500, mutation_rate: float = 0.1):
    print(f"Iniciando Algoritmo Genético para {n_queens}-Reinas...")
    start_time = time.time()
    
    # Generación de la población inicial
    population = generate_initial_population(population_size, n_queens)
    
    generation = 0
    while True:
        # Cálculo de la función de fitness
        fitness_scores = [calculate_fitness(ind) for ind in population]
        
        # Se verifica si se encontró una solución perfecta
        max_fitness = max(fitness_scores)
        best_idx = fitness_scores.index(max_fitness)
        best_individual = population[best_idx]
        
        if max_fitness == 1.0:
            end_time = time.time()
            
            print(f"\n¡Solución perfecta encontrada en la generación {generation}!")
            print(f"Tiempo de ejecución: {end_time - start_time:.4f} segundos")
            visualize_board(best_individual)
            return best_individual
        
        # Selección de los individuos mas aptos
        parents = select_fitted_population(population, fitness_scores)
        
        next_generation = [best_individual] # Elitismo: Preservar siempre al mejor de la generación
        
        # Generación de la siguiente generación
        while len(next_generation) < population_size:
            # Selección de dos padres aleatorios del pool de padres seleccionados
            parent1, parent2 = random.sample(parents, 2)
            
            # Cruce
            offspring1, offspring2 = crossover(parent1, parent2)
            
            # Mutación
            offspring1 = mutate(offspring1, mutation_rate)
            offspring2 = mutate(offspring2, mutation_rate)
            
            next_generation.extend([offspring1, offspring2])
            
        # Para asegurar el tamaño exacto de la población
        population = next_generation[:population_size]
        generation += 1
        
        # Si no hay solución después de un múltiplo de 500 generaciones, inyectamos diversidad genética
        if generation % 500 == 0:
            print(f"Generación {generation} alcanzada, mejor aptitud hasta ahora: {max_fitness:.4f}. Inyectando nuevo material genético...")
            # Extinción Masiva: Para evitar convergencia prematura por clones, mantenemos SOLO al mejor individuo
            # y reemplazamos a todos los demás con sangre completamente nueva.
            best_idx = fitness_scores.index(max_fitness)
            best_ind = population[best_idx]
            fresh_population = generate_initial_population(population_size - 1, n_queens)
            population = [best_ind] + fresh_population

if __name__ == "__main__":
    run_genetic_algorithm()

File: genetic_algorithm.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define N_QUEENS 32
#define POP_SIZE 2000
#define MUTATION_RATE 0.5

// Función auxiliar para simular índices de arreglos negativos como en Python
int get_python_index(int idx, int max_size) {
    if (idx < 0) return max_size + idx;
    return idx;
}

// Mezcla un arreglo aleatoriamente (Fisher-Yates) para generar permutaciones válidas
void shuffle(int *array, int n) {
    for (int i = n - 1; i > 0; i--) {
        int j = rand() % (i + 1);
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}

// Genera la población inicial usando PERMUTACIONES exclusivas
void generate_initial_population(int pop[POP_SIZE][N_QUEENS]) {
    for (int i = 0; i < POP_SIZE; i++) {
        for (int j = 0; j < N_QUEENS; j++) {
            pop[i][j] = j;
        }
        shuffle(pop[i], N_QUEENS);
    }
}

// Calcula el puntaje de aptitud (fitness score)
// Optimizada: Ya no calcula colisiones horizontales (que son 0 por defecto).
double calculate_fitness(int individual[N_QUEENS]) {
    double max_fitness = (N_QUEENS * (N_QUEENS - 1)) / 2.0;
    
    int left_diagonal[2 * N_QUEENS];
    int right_diagonal[2 * N_QUEENS];
    for (int i = 0; i < 2 * N_QUEENS; i++) { 
        left_diagonal[i] = 0; 
        right_diagonal[i] = 0; 
    }
    
    for (int i = 0; i < N_QUEENS; i++) {
        int l_idx = get_python_index(i + individual[i] - 1, 2 * N_QUEENS);
        int r_idx = get_python_index(N_QUEENS - i + individual[i] - 2, 2 * N_QUEENS);
        left_diagonal[l_idx]++;
        right_diagonal[r_idx]++;
    }
    
    double diagonal_collisions = 0;
    for (int i = 0; i < 2 * N_QUEENS - 1; i++) {
        double counter = 0;
        if (left_diagonal[i] > 1) counter += left_diagonal[i] - 1;
        if (right_diagonal[i] > 1) counter += right_diagonal[i] - 1;
        
        int denom = N_QUEENS - abs(i - N_QUEENS + 1);
        if (denom != 0) {
            diagonal_collisions += counter / denom;
        }
    }
    
    double fitness_val = max_fitness - diagonal_collisions;
    return fitness_val / max_fitness;
}

// Selecciona los individuos usando Torneo (Tournament Selection)
void select_fitted_population(int pop[POP_SIZE][N_QUEENS], double fitness[POP_SIZE], int parents[POP_SIZE][N_QUEENS]) {
    int tournament_size = 5;
    for (int i = 0; i < POP_SIZE; i++) {
        int best_idx = rand() % POP_SIZE;
        for (int j = 0; j < tournament_size - 1; j++) {
            int idx = rand() % POP_SIZE;
            if (fitness[idx] > fitness[best_idx]) {
                best_idx = idx;
            }
        }
        for (int k = 0; k < N_QUEENS; k++) {
            parents[i][k] = pop[best_idx][k];
        }
    }
}

// Realiza el Cruce de Orden (Order Crossover - OX1)
void crossover(int parent1[N_QUEENS], int parent2[N_QUEENS], int off1[N_QUEENS], int off2[N_QUEENS]) {
    // Puntos de corte aleatorios
    int p1 = rand() % N_QUEENS;
    int p2 = rand() % N_QUEENS;
    int start = p1 < p2 ? p1 : p2;
    int end = p1 > p2 ? p1 : p2;
    
    // Arrays para seguimiento (valores de 0 a N-1)
    int in_off1[N_QUEENS] = {0};
    int in_off2[N_QUEENS] = {0};
    
    for(int i=0; i<N_QUEENS; i++) {
        off1[i] = -1;
        off2[i] = -1;
    }
    
    // Paso 1: Copiar sub-secuencia del padre respectivo
    for (int i = start; i <= end; i++) {
        off1[i] = parent1[i];
        in_off1[parent1[i]] = 1;
        
        off2[i] = parent2[i];
        in_off2[parent2[i]] = 1;
    }
    
    // Paso 2: Rellenar circularmente desde el final del corte
    int p2_idx1 = (end + 1) % N_QUEENS;
    int off_idx1 = (end + 1) % N_QUEENS;
    
    while (1) {
        int done = 1;
        for(int i=0; i<N_QUEENS; i++) {
            if(off1[i] == -1) { done = 0; break; }
        }
        if(done) break;
        
        int gene = parent2[p2_idx1];
        if (!in_off1[gene]) {
            off1[off_idx1] = gene;
            in_off1[gene] = 1;
            off_idx1 = (off_idx1 + 1) % N_QUEENS;
        }
        p2_idx1 = (p2_idx1 + 1) % N_QUEENS;
    }
    
    int p2_idx2 = (end + 1) % N_QUEENS;
    int off_idx2 = (end + 1) % N_QUEENS;
    
    while (1) {
        int done = 1;
        for(int i=0; i<N_QUEENS; i++) {
            if(off2[i] == -1) { done = 0; break; }
        }
        if(done) break;
        
        int gene = parent1[p2_idx2];
        if (!in_off2[gene]) {
            off2[off_idx2] = gene;
            in_off2[gene] = 1;
            off_idx2 = (off_idx2 + 1) % N_QUEENS;
        }
        p2_idx2 = (p2_idx2 + 1) % N_QUEENS;
    }
}

// Mutación por Inversión (Inversion Mutation)
void mutate(int ind[N_QUEENS]) {
    double r = (double)rand() / RAND_MAX;
    if (r <= MUTATION_RATE) {
        int p1 = rand() % N_QUEENS;
        int p2 = rand() % N_QUEENS;
        int start = p1 < p2 ? p1 : p2;
        int end = p1 > p2 ? p1 : p2;
        
        while (start < end) {
            int temp = ind[start];
            ind[start] = ind[end];
            ind[end] = temp;
            start++;
            end--;
        }
    }
}

// Visualiza el tablero
void visualize_board(int ind[N_QUEENS]) {
    printf("\nSolución encontrada: [");
    for(int i = 0; i < N_QUEENS; i++) {
        printf("%d%s", ind[i], i == N_QUEENS - 1 ? "]\n" : ", ");
    }
    
    for (int row = 0; row < N_QUEENS; row++) {
        for (int col = 0; col < N_QUEENS; col++) {
            if (ind[col] == row) printf("Q ");
            else printf(". ");
        }
        printf("\n");
    }
    printf("\n");
}

int main() {
    srand(time(NULL));
    
    // static para evitar Stack Overflow con poblaciones grandes
    static int population[POP_SIZE][N_QUEENS];
    static double fitness_scores[POP_SIZE];
    static int parents[POP_SIZE][N_QUEENS];
    static int next_generation[POP_SIZE][N_QUEENS];
    
    printf("Iniciando Algoritmo Genético para %d-Reinas...\n", N_QUEENS);
    clock_t start_time = clock();
    
    generate_initial_population(population);
    
    int generation = 0;
    while (1) {
        double max_fitness = -1.0;
        int best_idx = -1;
        
        for (int i = 0; i < POP_SIZE; i++) {
            fitness_scores[i] = calculate_fitness(population[i]);
            if (fitness_scores[i] > max_fitness) {
                max_fitness = fitness_scores[i];
                best_idx = i;
            }
        }
        
        if (max_fitness >= 1.0) {
            clock_t end_time = clock();
            double time_spent = (double)(end_time - start_time) / CLOCKS_PER_SEC;
            
            printf("\n¡Solución perfecta encontrada en la generación %d!\n", generation);
            printf("Tiempo de ejecución: %.4f segundos\n", time_spent);
            visualize_board(population[best_idx]);
            break;
        }
        
        select_fitted_population(population, fitness_scores, parents);
        
        // Elitismo: Mantener al mejor individuo en la pos 0
        for (int k = 0; k < N_QUEENS; k++) {
            next_generation[0][k] = population[best_idx][k];
        }
        
        int next_idx = 1;
        while (next_idx < POP_SIZE) {
            int p1_idx = rand() % POP_SIZE;
            int p2_idx = rand() % POP_SIZE;
            
            int off1[N_QUEENS], off2[N_QUEENS];
            crossover(parents[p1_idx], parents[p2_idx], off1, off2);
            
            mutate(off1);
            mutate(off2);
            
            for (int k = 0; k < N_QUEENS; k++) {
                next_generation[next_idx][k] = off1[k];
            }
            next_idx++;
            
            if (next_idx < POP_SIZE) {
                for (int k = 0; k < N_QUEENS; k++) {
                    next_generation[next_idx][k] = off2[k];
                }
                next_idx++;
            }
        }
        
        for (int i = 0; i < POP_SIZE; i++) {
            for (int j = 0; j < N_QUEENS; j++) {
                population[i][j] = next_generation[i][j];
            }
        }
        
        generation++;
        
        // Extinción Masiva: Prevenir estancamiento en óptimos locales
        if (generation % 500 == 0) {
            printf("Generación %d alcanzada, mejor aptitud hasta ahora: %.4f. Inyectando nuevo material genético...\n", generation, max_fitness);
            
            int best_ind[N_QUEENS];
            for (int k = 0; k < N_QUEENS; k++) {
                best_ind[k] = population[best_idx][k];
            }
            
            generate_initial_population(population);
            
            for (int k = 0; k < N_QUEENS; k++) {
                population[0][k] = best_ind[k];
            }
        }
    }
    
    return 0;
}
